src/toolchain.ml

```diff
@@ -0,0 +1,260 @@
+      ProgressOptions.create ~location:(`ProgressLocation Window)
+        ~title:"Installing Platform Tools (this might take a few minutes)"
+        ~cancellable:false ()
+    in
+    let task ~progress:_ ~token:_ =
+      let* is_installed = is_sandbox_installed t in
+      let* _ =
+        match is_installed with
+        | false -> setup_toolchain_sandbox t.project_sandbox
+        | true -> Promise.return (Ok ())
+      in
+      let cmd = get_install_command t tools in
+      let+ result = Cmd.output cmd in
+      match result with
+      | Ok _ ->
+        show_message `Info "The platform tools have been successfully installed";
+        Ojs.null
+      | Error err ->
+        show_message `Error "The installation of the Platform tools failed: %s"
+          err;
+        Ojs.null
+    in
+    let+ (_ : Ojs.t) =
+      Window.withProgress (module Ojs) ~options:progress_options ~task
+    in
+    ()
+
+let upgrade t =
+  let open Promise.Syntax in
+  let opam, switch = t.opam in
+  let progress_options =
+    ProgressOptions.create ~location:(`ProgressLocation Window)
+      ~title:"Cheking for Platform Tools updates" ~cancellable:false ()
+  in
+  let task ~progress:_ ~token:_ =
+    let* _ = Opam.update opam switch |> Cmd.output in
+    let+ result = Opam.upgrade opam switch |> Cmd.output in
+    match result with
+    | Ok _ -> Ojs.null
+    | Error err ->
+      show_message `Error "The upgrade of the Platform tools failed: %s" err;
+      ();
+      Ojs.null
+  in
+  let+ (_ : Ojs.t) =
+    Window.withProgress (module Ojs) ~options:progress_options ~task
+  in
+  ()
+
+let setup ?project_sandbox () =
+  let open Promise.Syntax in
+  let* compiler_version = compiler_version project_sandbox in
+  let sandbox_name = switch_of_version compiler_version in
+  let* opam_opt =
+    let open Promise.Option.Syntax in
+    let* opam = Opam.make () in
+    let+ switch = Opam.Switch.of_string sandbox_name |> Promise.return in
+    (opam, switch)
+  in
+  match opam_opt with
+  | None ->
+    let msg =
+      "Opam is required to install the toolchain. Install it manually to use \
+       the extension."
+    in
+    show_message `Error "%s" msg;
+    Promise.return (Error msg)
+  | Some ((opam, _) as opam_switch) ->
+    let t =
+      { opam = opam_switch; project_sandbox; version = compiler_version }
+    in
+    let (_ : unit Promise.t) =
+      let* result = opam_init opam in
+      match result with
+      | Error err ->
+        show_message `Error "Could not setup toolchain: %s" err;
+        Promise.return ()
+      | Ok () ->
+        let* is_installed = is_sandbox_installed t in
+        let* missing_tools =
+          match is_installed with
+          | false -> Tool.all t
+          | true ->
+            let* () = upgrade t in
+            Tool.detect_missing t
+        in
+        let+ () = install_missing_tools t missing_tools in
+        ()
+    in
+    Promise.return (Ok t)
+
+let to_pretty_string { version; _ } = version
```

Comment:
I think it would be better if the pretty string had "OCaml" at the beginning so it's more apparent what extension the status bar item is coming from.
```suggestion
let to_pretty_string { version; _ } = "OCaml " ^ version
```